File: 1079.letter-tile-possibilities/solution.cpp
class Solution {
public:
    int numTilePossibilities(string tiles) {
		return possibilities(tiles).size();
    }
private:
	unordered_set<string> possibilities(string tiles) {
		unordered_set<string> result;
		if (tiles.length() == 1) 
		{
			result.insert("" + tiles);
			return result;
		}
		for (int i = 0; i < tiles.length(); i++) {
			char ch = tiles[i];
			string str = tiles;
			str.erase(next(str.begin(), i));
			unordered_set<string> result2 = possibilities(str);
			for (string str2 : result2) {
				result.insert(str2 + ch);
			}
			result.insert(result2.begin(), result2.end());
		}
		return result;
	}
};

File: 1079.letter-tile-possibilities/solution.kt
class Solution {
    fun numTilePossibilities(tiles: String): Int {
        val m = HashMap<Char, Int>()
        for (c in tiles)
            m.compute(c) { _, v -> (v ?: 0) + 1 }
        val tile = m.keys

        fun dfs(i: Int): Int {
            if (i == 0) return 1
            var res = 1
            for (t in tile) {
                if (m[t]!! > 0) {
                    m.computeIfPresent(t) { _, v -> v - 1 }
                    res += dfs(i - 1)
                    m.computeIfPresent(t) { _, v -> v + 1 }
                }
            }
            return res
        }

        return dfs(tiles.length) - 1
    }
}
